src/openrct2-ui/scripting/CustomWindow.cpp

```diff
@@ -0,0 +1,793 @@
+        {
+            window = window_create(
+                { desc.X.value(), desc.Y.value() }, desc.Width, desc.Height, &window_custom_events, WC_CUSTOM, windowFlags);
+        }
+        else
+        {
+            window = window_create_auto_pos(desc.Width, desc.Height, &window_custom_events, WC_CUSTOM, windowFlags);
+        }
+
+        window->number = GetNewWindowNumber();
+        window->custom_info = new CustomWindowInfo(owner, desc);
+        window->enabled_widgets = (1 << WIDX_CLOSE);
+
+        // Set window colours
+        window->colours[0] = COLOUR_GREY;
+        window->colours[1] = COLOUR_GREY;
+        window->colours[2] = COLOUR_GREY;
+        auto numColours = std::min(std::size(window->colours), std::size(desc.Colours));
+        for (size_t i = 0; i < numColours; i++)
+        {
+            window->colours[i] = desc.Colours[i];
+        }
+
+        if (desc.IsResizable())
+        {
+            window->min_width = desc.MinWidth.value_or(0);
+            window->min_height = desc.MinHeight.value_or(0);
+            window->max_width = desc.MaxWidth.value_or(std::numeric_limits<uint16_t>::max());
+            window->max_height = desc.MaxHeight.value_or(std::numeric_limits<uint16_t>::max());
+        }
+        RefreshWidgets(window);
+        window_init_scroll_widgets(window);
+        window_custom_update_viewport(window);
+        return window;
+    }
+
+    static void window_custom_close(rct_window* w)
+    {
+        delete static_cast<CustomWindowInfo*>(w->custom_info);
+        w->custom_info = nullptr;
+    }
+
+    static void window_custom_mouseup(rct_window* w, rct_widgetindex widgetIndex)
+    {
+        switch (widgetIndex)
+        {
+            case WIDX_CLOSE:
+                window_close(w);
+                break;
+            default:
+            {
+                const auto& info = GetInfo(w);
+                const auto widgetDesc = info.GetCustomWidgetDesc(widgetIndex);
+                if (widgetDesc != nullptr)
+                {
+                    if (widgetDesc->Type == "button")
+                    {
+                        InvokeEventHandler(info.Owner, widgetDesc->OnClick);
+                    }
+                    else if (widgetDesc->Type == "checkbox")
+                    {
+                        auto& widget = w->widgets[widgetIndex];
+                        widget.flags ^= WIDGET_FLAGS::IS_PRESSED;
+                        bool isChecked = widget.flags & WIDGET_FLAGS::IS_PRESSED;
+
+                        widget_set_checkbox_value(w, widgetIndex, isChecked);
+
+                        std::vector<DukValue> args;
+                        auto ctx = widgetDesc->OnChange.context();
+                        duk_push_boolean(ctx, isChecked);
+                        args.push_back(DukValue::take_from_stack(ctx));
+                        InvokeEventHandler(info.Owner, widgetDesc->OnChange, args);
+                    }
+                }
+                break;
```

Comment:
```suggestion
```
We don't really need a `break` on the `default`

Reply:
I think some compilers require it.